lhotse/dataset/speech_synthesis.py

```diff
@@ -15,38 +16,66 @@ class SpeechSynthesisDataset(torch.utils.data.Dataset):
     .. code-block::
 
         {
-            'audio': (B x NumSamples) tensor
-            'features': (B x NumFrames x NumFeatures) tensor
-            'tokens': list of lists of characters (str)
+            'audio': (B x NumSamples) float tensor
+            'features': (B x NumFrames x NumFeatures) float tensor
+            'tokens': (B x NumTokens) int tensor
+            'features_lens': (B x NumFrames) int tensor
+            'tokens_lens': (B x NumTokens) int tensor
+            'sentences': list of str
         }
     """
 
-    def __init__(self, cuts: CutSet) -> None:
+    def __init__(
+        self,
+        cuts: CutSet,
+        collate_features: Callable[CutSet, [Tensor, Tensor]] = None,
+        collate_tokens: Callable[CutSet, [Tensor, Tensor]] = None,
+    ) -> None:
         super().__init__()
+
         validate(cuts)
+        for cut in cuts:
+            assert (
+                len(cut.supervisions) == 1
+            ), "Only the Cuts with single supervision are supported."
+
         self.cuts = cuts
 
-        # generate tokens from text
-        self.cut_id_to_token = {}
-        self.tokens = set()
-        for cut in cuts:
-            assert len(cut.supervisions) == 1, 'Only the Cuts with single supervision are supported.'
-            characters = list(cut.supervisions[0].text)
-            self.tokens.update(set(characters))
-            self.cut_id_to_token[cut.id] = characters
-        self.tokens = sorted(list(self.tokens))
+        self.collate_tokens = (
+            CollateTokens(unique_tokens(self.cuts))
```

Comment:
I don't understand how it manages to preserve the mapping between token-sequences and feature-sequences, with this unique_tokens() call.

Reply:
The initialization creates a mapping from tokens to integers based on all used tokens in the dataset. This mapping is then used to map tokens loaded in the dataset to integers through the CollateTokens.__call__ method. Or maybe I do not understand the question correctly. Could you try to re-iterate?